include/seastar/coroutine/async_generator.hh

```diff
@@ -0,0 +1,363 @@
+            return _consumer;
+        }
+        void await_resume() noexcept {}
+    };
+
+    yield_awaiter do_yield() noexcept {
+        return yield_awaiter{this, _consumer};
+    }
+
+public:
+    async_generator_promise_base() noexcept = default;
+    async_generator_promise_base(const async_generator_promise_base &) = delete;
+    async_generator_promise_base& operator=(const async_generator_promise_base &) = delete;
+    async_generator_promise_base(async_generator_promise_base &&) noexcept = default;
+    async_generator_promise_base& operator=(async_generator_promise_base &&) noexcept = default;
+
+    // lazily-started coroutine, do not execute the coroutine until
+    // the coroutine is awaited.
+    std::suspend_always initial_suspend() const noexcept {
+        return {};
+    }
+
+    yield_awaiter final_suspend() noexcept {
+        _value = nullptr;
+        return do_yield();
+    }
+
+    void unhandled_exception() noexcept {
+        _exception = std::current_exception();
+    }
+
+    void return_void() noexcept {}
+
+    // @return if the generator has reached the end of the sequence
+    bool finished() const noexcept {
+        return _value == nullptr;
+    }
+
+    void rethrow_if_unhandled_exception() {
+        if (_exception) {
+            std::rethrow_exception(std::move(_exception));
+        }
+    }
+
+    void run_and_dispose() noexcept final {
+        using handle_type = std::coroutine_handle<async_generator_promise_base>;
+        handle_type::from_promise(*this).resume();
+    }
+
+    seastar::task* waiting_task() noexcept final {
+        return _waiting_task;
+    }
+
+    class element_awaiter {
+        std::remove_cvref_t<Yielded> _value;
+        constexpr bool await_ready() const noexcept {
+            return false;
+        }
+        template <typename Promise>
+        constexpr void await_suspend(std::coroutine_handle<Promise> producer) noexcept {
+            auto& current = producer.promise();
+            producer._value = std::addressof(_value);
+        }
+        constexpr void await_resume() const noexcept {}
+    };
+
+private:
+    friend class next_awaiter<Yielded>;
+};
+
+template <typename Yielded>
+class next_awaiter {
+protected:
+    async_generator_promise_base<Yielded>* _promise = nullptr;
+    std::coroutine_handle<> _producer = nullptr;
+
+    explicit next_awaiter(std::nullptr_t) noexcept {}
+    next_awaiter(async_generator_promise_base<Yielded>& promise,
+                 std::coroutine_handle<> producer) noexcept
+        : _promise{std::addressof(promise)}
+        , _producer{producer} {}
+
+public:
+    bool await_ready() const noexcept {
+        return false;
+    }
+
+    template <typename Promise>
+    std::coroutine_handle<> await_suspend(std::coroutine_handle<Promise> consumer) noexcept {
+        _promise->_consumer = consumer;
+        return _producer;
+    }
+};
+
+} // namespace internal
+
+template<typename Ref, typename Value>
```

Comment:
What is Ref and what is Value?

Reply:
Why do we need both?

Reply:
will revise the doxygen document to explain the rationales behind this.

Reply:
also added a test to exercise this feature.